core/src/main/java/org/apache/iceberg/util/ExpireSnapshotUtil.java

```diff
@@ -0,0 +1,397 @@
+      .select("manifest_path", "added_snapshot_id", "deleted_data_files_count");
+
+  private static CloseableIterable<ManifestFile> readManifestFiles(
+      Snapshot snapshot, TableOperations ops) {
+
+    if (snapshot.manifestListLocation() != null) {
+      return Avro.read(ops.io().newInputFile(snapshot.manifestListLocation()))
+          .rename("manifest_file", GenericManifestFile.class.getName())
+          .classLoader(GenericManifestFile.class.getClassLoader())
+          .project(MANIFEST_PROJECTION)
+          .reuseContainers(true)
+          .build();
+
+    } else {
+      return CloseableIterable.withNoopClose(snapshot.allManifests());
+    }
+  }
+
+  /**
+   * Determines the manifest files which need to be inspected because they refer to data files which
+   * can be removed after a Snapshot Expiration.
+   *
+   * @param currentTableSnapshots A list of Snapshots Currently used by the Table
+   * @param validIds              The Ids of the Snapshots which have not been expired
+   * @param expiredIds            The Ids of the Snapshots which have been expired
+   * @param currentTableMetadata  The metadata of the table being expired
+   * @param ops                   The Table Operations module for the table in question, required
+   *                              for several IO operations
+   * @return
+   */
+  public static ManifestExpirationChanges determineManifestChangesFromSnapshotExpiration(
+      List<Snapshot> currentTableSnapshots, Set<Long> validIds, Set<Long> expiredIds,
+      TableMetadata currentTableMetadata, TableOperations ops) {
+
+    Set<ManifestFile> validManifests = getValidManifests(currentTableSnapshots, ops);
+    Set<ManifestFile> manifestsToScan = validManifestsInExpiredSnapshots(validIds,
+        currentTableMetadata);
+
+    List<Snapshot> snapshotsNotChangingTableState = filterOutSnapshotsInTableState(validIds,
+        currentTableMetadata);
+
+    // find manifests to clean up that were only referenced by snapshots that have expired
+    ManifestExpirationChanges manifestExpirationChanges =
+        findExpiredManifestsInUnusedSnapshots(snapshotsNotChangingTableState, validManifests,
+            currentTableMetadata, expiredIds, ops);
+
+    manifestExpirationChanges.getManifestsToScan().addAll(manifestsToScan);
+    return manifestExpirationChanges;
+  }
+
+  public static SnapshotExpirationChanges getExpiredSnapshots(
+      TableOperations ops, TableMetadata originalMetadata) {
+
+    TableMetadata current = ops.refresh();
+
+    Set<Long> validIds = Sets.newHashSet();
+    for (Snapshot snapshot : current.snapshots()) {
+      validIds.add(snapshot.snapshotId());
+    }
+
+    Set<Long> expiredIds = Sets.newHashSet();
+    for (Snapshot snapshot : originalMetadata.snapshots()) {
+      long snapshotId = snapshot.snapshotId();
+      if (!validIds.contains(snapshotId)) {
+        // the snapshot was expired
+        LOG.info("Expired snapshot: {}", snapshot);
```

Comment:
I think it would make sense to log this entry in callers such as `RemoveSnapshots` rather than in the utility. For example, we could log a list of expired snapshots after this method is called.

Reply:
sgtm

Reply:
The only difficulty here is that we don't save the actual snapshots in this method, I can change the wrapper to save actual Snapshot objects instead of just their Ids? This would require a bit more refactoring